tools/job-cache/job-cache.cpp

```diff
@@ -478,17 +707,103 @@ class Cache {
         input_dirs(db),
         output_files(db),
         transact(db),
+        matching_jobs(db),
         dir(std::move(_dir)),
         rng(wcl::xoshiro_256::get_rng_seed()) {}
 
+  // TODO: Unlike reading, we need to account for
+  // directory remappings here.
+  wcl::optional<MatchingJob> read(const FindJobRequest &find_request) {
+    wcl::optional<MatchingJob> result;
+
+    // We run the matching job in a transaction. This ensures
+    // that we get the *complete* set of output files. This
+    // allows us to know if one of the job files was tampered
+    // with while we're copying files into place.
+    transact.run([this, &find_request, &result]() { result = matching_jobs.find(find_request); });
+
+    // Return early if there was no match.
+    if (!result) return {};
+
+    // We need a tmp directory to put these outputs into
+    std::string tmp_job_dir = dir + "/tmp_outputs_" + rng.unique_name();
+    mkdir_no_fail(tmp_job_dir.c_str());
+
+    // We also need to know what directory we're reading out of
+    std::string job_dir = dir + "/" + std::to_string(result->job_id & 0xFF) + "/" +
```

Comment:
nit: looks like a good candiate for `join`?

Reply:
not quite because its not an array of values.